servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AbstractProcessorBuffer.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright © 2020 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.concurrent.api;
+
+import io.servicetalk.concurrent.internal.TerminalNotification;
+
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.concurrent.internal.TerminalNotification.complete;
+import static io.servicetalk.concurrent.internal.TerminalNotification.error;
+import static java.util.concurrent.atomic.AtomicReferenceFieldUpdater.newUpdater;
+
+abstract class AbstractProcessorBuffer<T> implements ProcessorBuffer<T> {
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<AbstractProcessorBuffer,
+            TerminalNotification> terminalUpdater = newUpdater(AbstractProcessorBuffer.class,
+            TerminalNotification.class, "terminal");
+    private static final Object NULL_ITEM = new Object();
+
+    @Nullable
+    private volatile TerminalNotification terminal;
+
+    @Override
+    public final void add(@Nullable final T item) {
+        final TerminalNotification terminalNotification = terminal;
+        if (terminalNotification != null) {
+            throw new IllegalStateException("Buffer " + this + " is already terminated: " + terminal);
+        }
+        addItem(item == null ? NULL_ITEM : item);
+    }
+
+    @Override
+    public final void terminate() {
+        if (terminalUpdater.compareAndSet(this, null, complete())) {
```

Comment:
Consider storing the return value of `complete()` and use the same value for `addTerminal(...)`. Just to be sure its the same instance at the end...